### ISSUE TITLE
Raise precedence of xdg-open in default rifle.conf

### ISSUE THREAD

**Issue by tmewett** [ranger/ranger#2225]
#### Runtime Environment

- Ranger version/commit: ranger-master (Arch community package 1.9.3.143.g5047d9e5-1)

#### Current Behavior
Currently, the Rifle rules for xdg-open and open - command-line programs which use the typical "graphical" MIME configuration - are very low down in the default configuration.

#### Proposed Behaviour
I think it would be a better default to have these generic openers *above* all graphical application for all file types.

#### Rationale

When graphical programs are installed, they typically add their associated file types into the MIME database of the system, via e.g. .desktop files on Linux. Pretty much any graphical program that needs to open a file uses this database and the user can change the defaults easily from a graphical file manager.

Users who don't exclusively use terminal FMs will have their MIME database configured to their preference, and those who do will have some default, which is no worse than a pre-defined 'guess' rifle.conf. Therefore, it stands to reason that in the absence of better information, the MIME database (via xdg-open) is the best guess to open a file with a graphical program.

The exception to this is for terminal applications, which typically do not come with .desktop files or MIME associations. Hence, these should continue to be higher-precedence than xdg-open.

Rifle rules for graphical programs can remain, as fallback in case xdg-open is not available.

My hope is this is a smarter default for new users.

I would be happy to PR this, just wanted to discuss and see if you agree! Thanks :)

**Comment by toonn** [comment]
I disagree, they're only fallback methods. There are tons of users who don't know how to configure those MIME type associations. For those that have `xdg-open` set up just the way they like it it's only a single rule to add to `rifle.conf`. The default rules also match things other than MIME types, I'm not sure `xdg-open` can do this? If we changed the order then `xdg-open` would probably open these files with a fallback method and rifle's rules would never be used.

**Comment by tmewett** [comment]
Thanks for the response!

> There are tons of users who don't know how to configure those MIME type associations

It's not necessary that users know. If a user has xdg-open installed but doesn't use graphical FMs, it only serves to give a sensible default based on installed apps - which is exactly what default rifle.conf does, hence my claim that it is "no worse." If they want to change the opener, they edit rifle.conf with their own settings, just like they have to do now.

Also, anyone who uses a graphical FM does know how to configure them - you right click a file and choose "Open with..." and set it as the default. There are also default app control panels in many DEs. That's what I mean by "users who don't exclusively use terminal FMs will have their MIME database already configured to their preference."

> For those that have xdg-open set up just the way they like it it's only a single rule to add to rifle.conf.

Indeed, but with this change it would be zero config necessary, without effectively changing the config requirements of other users.

>The default rules also match things other than MIME types, I'm not sure xdg-open can do this?

You're right, but from my reading, simple application associations (the type of rule I argue xdg-open should take priority over) effectively don't do more than what xdg-open. In addition to all rules for terminal apps, there are various others which would make sense to keep above it, such as the Wine rule. 

Would it be valuable to see some specifics? I could do some testing and make a draft PR

**Comment by toonn** [comment]
You're assuming people only install ranger in addition to a GUI file manager and on a DE. But ranger attracts exactly the kind of user who might only use a window manager and may be avoiding GUI file managers. The usual way this gets configured is when you install a program or use it to open a file it asks to be the default opener associated to some MIME types. Most of the programs in rifle.conf don't do this so they are unlikely to be configured to be used by `xdg-open` afaict.

A PR isn't necessary yet. The change would be simple enough.

**Comment by tmewett** [comment]
> You're assuming people only install ranger in addition to a GUI file manager and on a DE.

Could you explain where I'm assuming this? I think I've covered all cases.

> The usual way this gets configured is when you install a program or use it to open a file it asks to be the default opener associated to some MIME types

That's not how most applications make associations on Linux. In fact, I'm not sure know of any which do that (maybe some web browsers?) I'm not sure this is relevant to this change though, unless I'm wrong?

> Most of the programs in rifle.conf don't do this so they are unlikely to be configured to be used by xdg-open afaict.

Pretty much any graphical program/package that installs a .desktop file will have MIME associations in it. This includes quite a lot of the programs in Rifle's defaults rules, and these are the rules that I'm suggesting move below xdg-open. Terminal applications must stay above.
